x-pack/plugins/saved_objects_tagging/server/usage/schema.ts

```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { MakeSchemaFrom } from '../../../../../src/plugins/usage_collection/server';
+import { TaggingUsageData, ByTypeTaggingUsageData } from './types';
+
+const perTypeSchema: MakeSchemaFrom<ByTypeTaggingUsageData> = {
+  usedTags: { type: 'integer' },
+  taggedObjects: { type: 'integer' },
+};
+
+export const tagUsageCollectorSchema: MakeSchemaFrom<TaggingUsageData> = {
+  usedTags: { type: 'integer' },
+  taggedObjects: { type: 'integer' },
+
+  types: {
+    dashboard: perTypeSchema,
```

Comment:
we collect tags usage data for more than just these two saved object types. I think we should try to avoid having this kind of dynamic fields in our schema and it will also increase the field count on our telemetry cluster.

What about:
```
perSavedObjectType: {
  usedTags: { type: 'integer' },
  taggedObjects: { type: 'integer' },
  savedObjectType: {type: 'string'},
}
```

Reply:
> perSavedObjectType: { ...

Should we use a `nested` or `array` type in that case?

That would work. However I would be sure that is would be alright to the way telemetry consumes / uses the data.

When I look at the other collectors, none seems to be used nested fields, which is why I'm asking:

https://github.com/elastic/kibana/blob/d67a421e68508aed1413750c2e1696d0d9921f39/src/plugins/kibana_usage_collection/server/collectors/application_usage/schema.ts#L52-L59

note that this could only be a design flaw when the collector was implemented. @elastic/kibana-telemetry?

Reply:
Hey! This is a legacy issue that no longer applies: in the past, before having the schema implementation, the indexer in the remote service indexed the data in the same format it received it from Kibana. This meant that we had to keep the mappings in mind when developing the collectors (and `nested` was never an option because Kibana does not natively support viz with that type yet).
New collectors/fields shouldn't have those contraints anymore because the new indexer actually reshapes the payloads and splits the info into multiple indices based on the different needs to consume the information by our PMs and Analysts.
If we foresee the number of keys growing dynamically, I think the `array` approach would be best.

cc @mindbat so he can share his thoughts from the indexer-maintainer POV 🙂

Reply:
> New collectors/fields shouldn't have those contraints anymore because the new indexer actually reshapes the payloads and splits the info into multiple indices based on the different needs to consume the information by our PMs and Analysts.

@afharo:  We're not sending Kibana data to Telemetry Next _yet_ and the legacy indexer doesn't reshape data by default. We would still need to update the indexer to consume the new data, [reshape it]((https://github.com/elastic/telemetry/blob/53aa5a17e6e80b600e217b2f158a69d800952ca1/src/indexer/channel_handlers/legacy_telemetry/index.ts#L34-L40)) and index it into a custom extended index.

Reply:
So should we stick with the static, per type, mappings for now?

Reply:
I agree with @TinaHeiligers: unless your team is willing to be an early adopter of Telemetry-Next (which is still _very_ much beta), I recommend adopting a schema that fits within the constraints of the current system (which I believe means static mappings, no nested fields).

Reply:
Even if we just index data as-is, can you elaborate why we couldn't use an array or nested field type for the mappings?

Reply:
> and nested was never an option because Kibana does not natively support viz with that type yet

I missed @afharo's comment and I assume this applies to array fields too https://github.com/elastic/kibana/issues/58175

It makes sense that we want to do some processing inside Telemetry Next, but I think it would be much more flexible if Kibana had more control around which documents are created. If telemetry is uploaded as NDJSON then usage collectors could return several documents instead of always having a single large payload that needs to be handled and split by the server. This is probably a big effort, but maintaining a static schema for dynamic data like applications and saved object types that change every release feels like a maintenance nightmare.

So I guess for this PR we just need to add all the known saved object types to the mappings.

Reply:
> So I guess for this PR we just need to add all the known saved object types to the mappings.

We only need to add types that currently supports the tagging feature. That would just be dashboard, visualizations and maps. Will add maps. Further additions should only be done when the app/type implements the tagging feature.

Reply:
Done in https://github.com/elastic/kibana/pull/83160/commits/7b6d76ef58a7fc0b7a990c303f135cb76fbe1988: added `map` to the schema, and updated the readme to state that developers should update the schema when adding tagging support to additional SO types.

Reply:
@rudolf 

> Even if we just index data as-is, can you elaborate why we couldn't use an array field type for the mappings?

Without any transformations on the remote service indexer, we'd end up with data mapped as an array in elasticsearch and face the usual [problem](https://www.elastic.co/guide/en/elasticsearch/reference/current/array.html) of not being able to meaningfully query or visualize that data. It's a situation we're trying to fix with `ui_metric` right now, for example.